File: Geometry/GrahamScan.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef struct point {
  double x;
  double y;
} Point;

// A function to compare two points based on their polar angle
int compare(const void* vp1, const void* vp2) {
  Point* p1 = (Point*) vp1;
  Point* p2 = (Point*) vp2;
  double angle1 = atan2(p1->y, p1->x);
  double angle2 = atan2(p2->y, p2->x);
  if (angle1 < angle2) return -1;
  else if (angle1 > angle2) return 1;
  else return 0;
}

// A function to check if three points make a left turn
int left_turn(Point p1, Point p2, Point p3) {
  double cross_product = (p2.x - p1.x) * (p3.y - p1.y) - (p2.y - p1.y) * (p3.x - p1.x);
  return (cross_product > 0);
}

// A function to compute the convex hull of a set of points using the Graham scan algorithm
void convex_hull(Point* points, int num_points) {
  // Sort the points based on their polar angle
  qsort(points, num_points, sizeof(Point), compare);

  // Create a stack to hold the points of the convex hull
  Point* stack = (Point*) malloc(num_points * sizeof(Point));
  int top = 2;
  stack[0] = points[0];
  stack[1] = points[1];

  // Iterate through the remaining points
  for (int i = 2; i < num_points; i++) {
    // Pop points from the stack while they make a right turn with the next point
    while (top > 0 && !left_turn(stack[top-2], stack[top-1], points[i])) {
      top--;
    }
    // Push the next point onto the stack
    stack[top] = points[i];
    top++;
  }

  // Print the points of the convex hull
  printf("Convex Hull:\n");
  for (int i = 0; i < top; i++) {
    printf("(%lf, %lf)\n", stack[i].x, stack[i].y);
  }

  // Free the memory allocated for the stack
  free(stack);
}

int main() {
  // Example usage
  Point points[] = { {0, 0}, {0, 1}, {1, 0}, {1, 1}, {0.5, 0.5} };
  int num_points = sizeof(points) / sizeof(Point);
  convex_hull(points, num_points);

  return 0;
}

File: Geometry/GrahamScan.java
import java.util.*;

class Point {
    double x;
    double y;

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }
}

public class GrahamScan {
    // A function to compare two points based on their polar angle
    static class PolarAngleComparator implements Comparator<Point> {
        public int compare(Point p1, Point p2) {
            double angle1 = Math.atan2(p1.y, p1.x);
            double angle2 = Math.atan2(p2.y, p2.x);
            if (angle1 < angle2) return -1;
            else if (angle1 > angle2) return 1;
            else return 0;
        }
    }

    // A function to check if three points make a left turn
    static boolean leftTurn(Point p1, Point p2, Point p3) {
        double crossProduct = (p2.x - p1.x) * (p3.y - p1.y) - (p2.y - p1.y) * (p3.x - p1.x);
        return (crossProduct > 0);
    }

    // A function to compute the convex hull of a set of points using the Graham scan algorithm
    static List<Point> convexHull(Point[] points) {
        // Sort the points based on their polar angle
        Arrays.sort(points, new PolarAngleComparator());

        // Create a stack to hold the points of the convex hull
        Stack<Point> stack = new Stack<Point>();
        stack.push(points[0]);
        stack.push(points[1]);

        // Iterate through the remaining points
        for (int i = 2; i < points.length; i++) {
            // Pop points from the stack while they make a right turn with the next point
            while (stack.size() > 1 && !leftTurn(stack.elementAt(stack.size()-2), stack.peek(), points[i])) {
                stack.pop();
            }
            // Push the next point onto the stack
            stack.push(points[i]);
        }

        // Convert the stack to a list and return it
        List<Point> convexHull = new ArrayList<Point>(stack);
        return convexHull;
    }

    public static void main(String[] args) {
        // Example usage
        Point[] points = { new Point(0, 0), new Point(0, 1), new Point(1, 0), new Point(1, 1), new Point(0.5, 0.5) };
        List<Point> convexHull = convexHull(points);
        System.out.println("Convex Hull:");
        for (Point p : convexHull) {
            System.out.println("(" + p.x + ", " + p.y + ")");
        }
    }
}
